packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +100,118 @@ class OptimisticOracleClient {
-            )
-            .call(),
-        {
-          concurrency: WEB3_CALLS_BATCH_SIZE
-        }
-      )
-    ]);
+    // Store price requests that have NOT been proposed to yet:
+    const unproposedPriceRequests = requestEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasProposal = proposalEvents.find(proposalEvent => this._getPriceRequestKey(proposalEvent) === key);
+      return hasProposal === undefined;
+    });
+    this.unproposedPriceRequests = unproposedPriceRequests.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        currency: event.returnValues.currency,
+        reward: event.returnValues.reward,
+        finalFee: event.returnValues.finalFee
+      };
+    });
 
-    // Sort price requests based on state.
-    requestStates.map((state, index) => {
-      let requestData = {
-        requester: requestEvents[index].returnValues.requester,
-        identifier: this.hexToUtf8(requestEvents[index].returnValues.identifier),
-        timestamp: requestEvents[index].returnValues.timestamp,
-        currency: requestEvents[index].returnValues.currency,
-        reward: requestEvents[index].returnValues.reward,
-        finalFee: requestEvents[index].returnValues.finalFee
+    // Store proposals that have NOT been disputed:
+    let undisputedProposals = proposalEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasDispute = disputeEvents.find(disputeEvent => this._getPriceRequestKey(disputeEvent) === key);
+      return hasDispute === undefined;
+    });
+    undisputedProposals = undisputedProposals.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        proposer: event.returnValues.proposer,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        proposedPrice: event.returnValues.proposedPrice,
+        expirationTimestamp: event.returnValues.expirationTimestamp
       };
-      if (state === OptimisticOracleRequestStatesEnum.REQUESTED) {
-        this.priceRequests.push(requestData);
-      } else if (state === OptimisticOracleRequestStatesEnum.PROPOSED) {
-        // TODO: Add data specific for proposals,
-        // or alternatively search for ProposedPrice events.
-        this.priceProposals.push(requestData);
-      }
     });
 
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that can be settled:
+    let resolvedPrices = await Promise.all(
```

Comment:
nit: `resolvedPrices` seems to be a misnomer here and below. It appears as if this array is actually being used to store events. Thoughts on renaming it?

Reply:
Good point, renamed to `resolvedDisputeEvents`